[custom_ops][perf] Move expensive pytree traversals of tensors to C++

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #148555

(benchmark for 1 call)

Before:
```
└─ $ python ~/task_custom_ops_perf/test_custom_ops_perf_repro.py
DO_BENCH mutate: 77.72445678710938 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/mutate.json
DO_BENCH no_mutate: 64.61143493652344 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/no_mutate.json
DO_BENCH direct_mutate: 11.682510375976562 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_mutate.json
DO_BENCH direct_no_mutate: 18.596649169921875 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_no_mutate.json
```

After:
```
└─ $ python ~/task_custom_ops_perf/test_custom_ops_perf_repro.py
DO_BENCH mutate: 47.6837158203125 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/mutate.json
DO_BENCH no_mutate: 31.709671020507812 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/no_mutate.json
DO_BENCH direct_mutate: 10.967254638671875 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_mutate.json
DO_BENCH direct_no_mutate: 10.728836059570312 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_no_mutate.json
```

IvanKobzarev committed:
[DEBUG] Custom ops perf

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[DEBUG] Custom ops perf"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[DEBUG] Custom ops perf"




[ghstack-poisoned]

zou3519 commented on test/test_custom_ops_perf_repro.py:
```diff
@@ -0,0 +1,147 @@
+
+
+@torch.library.custom_op("foo::baz", device_types="cpu", mutates_args=(["a"]))
+def baz(
```
Btw @albanD mentioned trying just looking for requires_grad=True in Python instead of C++ and benchmarking the two. Hypothesis is that for smaller number of tensors the Python version might be faster

albanD replied:
```python
In [1]: import torch
   ...: from torch.utils import _pytree
   ...: from torch import Tensor
   ...: 
   ...: a = [torch.rand(10, 10) for _ in range(10)]
   ...: 
   ...: %timeit _pytree.tree_any_only(Tensor, lambda x: x.requires_grad, a)
   ...: %timeit _pytree.tree_any_only(Tensor, lambda x: x.requires_grad, a)
   ...: 
   ...: %timeit any(t.requires_grad for t in a)
   ...: %timeit any(t.requires_grad for t in a)
   ...: 
   ...: a = [torch.rand(10, 10, requires_grad=True) for _ in range(10)]
   ...: 
   ...: %timeit _pytree.tree_any_only(Tensor, lambda x: x.requires_grad, a)
   ...: %timeit _pytree.tree_any_only(Tensor, lambda x: x.requires_grad, a)
   ...: 
   ...: %timeit any(t.requires_grad for t in a)
   ...: %timeit any(t.requires_grad for t in a)
   ...: 
   ...: 
16.2 µs ± 56.8 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
16.4 µs ± 137 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
2.07 µs ± 26.9 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
2.1 µs ± 50.6 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
5.03 µs ± 92.6 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
5.05 µs ± 123 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
795 ns ± 10.7 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
804 ns ± 14.3 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

IvanKobzarev replied:
@zou3519 @albanD 
Interesting, why there is so big difference for `requires_grad=True` and `requires_grad=False`? Is it because of additional wrapping into `Variable` ?


In our case the main overhead comes from additional pytree any(Tensor) logic which does two `isinstance`, `_is_leaf`, `not_list_of_tensor` checks etc. e.g. profiles of the baseline:

<img width="887" alt="Screenshot 2025-03-11 at 07 47 51" src="https://github.com/user-attachments/assets/376b01b7-a98a-446f-b60b-1b945029c9e1" />
where each iteration is: (2 builtin blocks are isinstance)
<img width="1307" alt="Screenshot 2025-03-11 at 07 48 25" src="https://github.com/user-attachments/assets/f08ac7f8-28f9-4054-809b-1cfc867a70fd" />

albanD replied:
> Interesting, why there is so big difference for requires_grad=True and requires_grad=False?

any() short circuits as soon as it finds the first "True" :D

IvanKobzarev committed:
Update on "[custom_ops][perf] Move expensive pytree traversals of tensors to C++"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[custom_ops][perf] Move expensive pytree traversals of tensors to C++"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[custom_ops][perf] Move expensive pytree traversals of tensors to C++"


(benchmark for 1 call)

Before:
```
└─ $ python ~/task_custom_ops_perf/test_custom_ops_perf_repro.py
DO_BENCH mutate: 77.72445678710938 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/mutate.json
DO_BENCH no_mutate: 64.61143493652344 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/no_mutate.json
DO_BENCH direct_mutate: 11.682510375976562 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_mutate.json
DO_BENCH direct_no_mutate: 18.596649169921875 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_no_mutate.json
```

After:
```
└─ $ python ~/task_custom_ops_perf/test_custom_ops_perf_repro.py
DO_BENCH mutate: 47.6837158203125 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/mutate.json
DO_BENCH no_mutate: 31.709671020507812 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/no_mutate.json
DO_BENCH direct_mutate: 10.967254638671875 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_mutate.json
DO_BENCH direct_no_mutate: 10.728836059570312 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_no_mutate.json
```


[ghstack-poisoned]

zou3519 commented on test/test_custom_ops.py:
```diff
@@ -3910,6 +3910,67 @@ def fvmap2(info, in_dims, x, y):
+            )
+        )
+        self.assertFalse(test_fn((x,), None, (2 * x,)))
+        self.assertTrue(
+            test_fn(
+                (),
+                {"a": x.view(-1)},
+                (x,),
+            )
+        )
```
Can you add a test case where the output is exactly the same as the input? e.g.
self.assertTrue(
            test_fn(
                (x,),
                {},
                (x,),
            )
        )

zou3519 commented on test/test_custom_ops.py:
```diff
@@ -3910,6 +3910,67 @@ def fvmap2(info, in_dims, x, y):
+                (torch.tensor([]),),
+            )
+        )
+
```
can you add some tests where the args have mixed Tensors and TensorList? e.g.

test_fn((x, [y, z]), ...)

For both this function and for test_any_requires_grad

zou3519 commented on torch/_library/utils.py:
```diff
@@ -373,6 +373,28 @@ def check_aliasing_constraint(name, prev, result, get_module=lambda: "???"):
 
+def _c_check_aliasing_constraint(name, args, kwargs, result, get_module=lambda: "???"):
+    """
+    custom operators' outputs must not have any aliases
```
can you document here that we only handle Tensors and List[Tensors] in args, kwargs (and nothing else?)

zou3519 commented on torch/csrc/autograd/init.cpp:
```diff
@@ -955,6 +955,120 @@ static PyObject* is_fwd_grad_enabled(PyObject* _unused, PyObject* arg) {
+// Args:
+// args[0] - inputs args
+// args[1] - inputs kwargs
+// args[2] - outputs
```
can you document here that we only handle Tensors and List[Tensors] in args, kwargs , outputs (other data sturctures of tensors, like List[List[Tensor]] will not be peeked into)

zou3519 commented on torch/csrc/autograd/init.cpp:
```diff
@@ -955,6 +955,120 @@ static PyObject* is_fwd_grad_enabled(PyObject* _unused, PyObject* arg) {
+    PyObject* args,
+    PyObject* kwargs) {
+  HANDLE_TH_ERRORS
+  bool has_requires_grad = false;
+  visit_tensors(args, kwargs, [&has_requires_grad](at::Tensor& t) {
+    if (t.requires_grad()) {
+      has_requires_grad = true;
+      return true;
+    }
+    return false;
+  });
```
Does this do an early return?

zou3519 replied:
Yeah it doesn't look like your code does an early return.

IvanKobzarev replied:
This shortcircuit does not fully shortcircuit for recursive impl, will correct handling in visit()

zou3519 commented on torch/csrc/autograd/init.cpp:
```diff
@@ -955,6 +955,120 @@ static PyObject* is_fwd_grad_enabled(PyObject* _unused, PyObject* arg) {
+  PyObject* inps = PyTuple_GET_ITEM(args, 0);
+  PyObject* inps_kwargs = PyTuple_GET_ITEM(args, 1);
+  PyObject* outs = PyTuple_GET_ITEM(args, 2);
+  std::unordered_set<void*> s;
```
we have a ska::hash_map data structure somewhere, could be interesting to play around with it. I'm not sure what's faster

zou3519 reviewed (commented):
this looks good to me. The only comment I have is that we should be doing an early return for any_requires_grad (if the first tensor requires grad then we should short-circuit).

zou3519 commented on torch/csrc/autograd/init.cpp:
```diff
@@ -955,6 +955,120 @@ static PyObject* is_fwd_grad_enabled(PyObject* _unused, PyObject* arg) {
+}
+
+template <bool skip_tensors_in_non_tensorlist = false>
+static void visit_tensors(
+    PyObject* args,
+    PyObject* kwargs,
+    const std::function<bool(at::Tensor&)>& visit_tensor) {
+  if (args && PyTuple_Check(args)) {
+    for (const auto i : c10::irange(PyTuple_GET_SIZE(args))) {
+      visit<skip_tensors_in_non_tensorlist>(
+          PyTuple_GET_ITEM(args, i), visit_tensor);
+    }
+  }
+  if (kwargs && PyDict_Check(kwargs)) {
+    auto vals = PyDict_Values(kwargs);
+    for (const auto i : c10::irange(PyList_GET_SIZE(vals))) {
+      visit<skip_tensors_in_non_tensorlist>(
+          PyList_GET_ITEM(vals, i), visit_tensor);
+    }
+  }
+}
```
@IvanKobzarev this function doesn't short circuit, right? `any_requires_grad` calls visit_tensors. This is the function that I would expect to short circuit.

IvanKobzarev committed:
Update on "[custom_ops][perf] Move expensive pytree traversals of tensors to C++"


(benchmark for 1 call)

Before:
```
└─ $ python ~/task_custom_ops_perf/test_custom_ops_perf_repro.py
DO_BENCH mutate: 77.72445678710938 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/mutate.json
DO_BENCH no_mutate: 64.61143493652344 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/no_mutate.json
DO_BENCH direct_mutate: 11.682510375976562 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_mutate.json
DO_BENCH direct_no_mutate: 18.596649169921875 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_no_mutate.json
```

After:
```
└─ $ python ~/task_custom_ops_perf/test_custom_ops_perf_repro.py
DO_BENCH mutate: 47.6837158203125 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/mutate.json
DO_BENCH no_mutate: 31.709671020507812 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/no_mutate.json
DO_BENCH direct_mutate: 10.967254638671875 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_mutate.json
DO_BENCH direct_no_mutate: 10.728836059570312 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_no_mutate.json
```


[ghstack-poisoned]

IvanKobzarev committed:
Update on "[custom_ops][perf] Move expensive pytree traversals of tensors to C++"


(benchmark for 1 call)

Before:
```
└─ $ python ~/task_custom_ops_perf/test_custom_ops_perf_repro.py
DO_BENCH mutate: 77.72445678710938 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/mutate.json
DO_BENCH no_mutate: 64.61143493652344 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/no_mutate.json
DO_BENCH direct_mutate: 11.682510375976562 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_mutate.json
DO_BENCH direct_no_mutate: 18.596649169921875 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_no_mutate.json
```

After:
```
└─ $ python ~/task_custom_ops_perf/test_custom_ops_perf_repro.py
DO_BENCH mutate: 47.6837158203125 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/mutate.json
DO_BENCH no_mutate: 31.709671020507812 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/no_mutate.json
DO_BENCH direct_mutate: 10.967254638671875 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_mutate.json
DO_BENCH direct_no_mutate: 10.728836059570312 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_no_mutate.json
```


[ghstack-poisoned]

zou3519 commented on torch/csrc/autograd/init.cpp:
```diff
@@ -955,6 +955,133 @@ static PyObject* is_fwd_grad_enabled(PyObject* _unused, PyObject* arg) {
+    PyObject* kwargs) {
+  HANDLE_TH_ERRORS
+  bool has_requires_grad = false;
+  visit_tensors<true>(args, kwargs, [&has_requires_grad](at::Tensor& t) {
```
actually... I think the TensorList behavior might be wrong, but your PR at least improves on the current state so let's ship it and worry about the rest later.

If you're interested can you try to make a custom_op that takes in a List[Tensor?] input, have it accept (x, [None, y]) as input (where y is the only tensor that requires_grad), and see what happens?

IvanKobzarev commented:
@pytorchbot merge

zou3519 commented on torch/csrc/autograd/init.cpp:
```diff
@@ -955,6 +955,133 @@ static PyObject* is_fwd_grad_enabled(PyObject* _unused, PyObject* arg) {
+    }
+  }
+  if (kwargs && PyDict_Check(kwargs)) {
+    auto vals = PyDict_Values(kwargs);
```
auto vals = THPObjectPtr{PyDict_Values(dict)}; should do the trick

IvanKobzarev committed:
Update on "[custom_ops][perf] Move expensive pytree traversals of tensors to C++"


(benchmark for 1 call)

Before:
```
└─ $ python ~/task_custom_ops_perf/test_custom_ops_perf_repro.py
DO_BENCH mutate: 77.72445678710938 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/mutate.json
DO_BENCH no_mutate: 64.61143493652344 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/no_mutate.json
DO_BENCH direct_mutate: 11.682510375976562 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_mutate.json
DO_BENCH direct_no_mutate: 18.596649169921875 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_no_mutate.json
```

After:
```
└─ $ python ~/task_custom_ops_perf/test_custom_ops_perf_repro.py
DO_BENCH mutate: 47.6837158203125 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/mutate.json
DO_BENCH no_mutate: 31.709671020507812 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/no_mutate.json
DO_BENCH direct_mutate: 10.967254638671875 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_mutate.json
DO_BENCH direct_no_mutate: 10.728836059570312 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_no_mutate.json
```


[ghstack-poisoned]

zou3519 commented on test/test_custom_ops.py:
```diff
@@ -3910,6 +3910,125 @@ def fvmap2(info, in_dims, x, y):
+        t = torch.randn(2, 2)
+        t_refcount = sys.getrefcount(t)
+        test_fn(t, a=t)
+        assert sys.getrefcount(t) == t_refcount
```
nit: self.assertEqual?

zou3519 reviewed (approved):
thank you

IvanKobzarev committed:
Update on "[custom_ops][perf] Move expensive pytree traversals of tensors to C++"


(benchmark for 1 call)

Before:
```
└─ $ python ~/task_custom_ops_perf/test_custom_ops_perf_repro.py
DO_BENCH mutate: 77.72445678710938 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/mutate.json
DO_BENCH no_mutate: 64.61143493652344 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/no_mutate.json
DO_BENCH direct_mutate: 11.682510375976562 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_mutate.json
DO_BENCH direct_no_mutate: 18.596649169921875 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_no_mutate.json
```

After:
```
└─ $ python ~/task_custom_ops_perf/test_custom_ops_perf_repro.py
DO_BENCH mutate: 47.6837158203125 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/mutate.json
DO_BENCH no_mutate: 31.709671020507812 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/no_mutate.json
DO_BENCH direct_mutate: 10.967254638671875 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_mutate.json
DO_BENCH direct_no_mutate: 10.728836059570312 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_no_mutate.json
```


[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot merge

IvanKobzarev committed:
Update on "[custom_ops][perf] Move expensive pytree traversals of tensors to C++"


(benchmark for 1 call)

Before:
```
└─ $ python ~/task_custom_ops_perf/test_custom_ops_perf_repro.py
DO_BENCH mutate: 77.72445678710938 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/mutate.json
DO_BENCH no_mutate: 64.61143493652344 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/no_mutate.json
DO_BENCH direct_mutate: 11.682510375976562 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_mutate.json
DO_BENCH direct_no_mutate: 18.596649169921875 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_no_mutate.json
```

After:
```
└─ $ python ~/task_custom_ops_perf/test_custom_ops_perf_repro.py
DO_BENCH mutate: 47.6837158203125 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/mutate.json
DO_BENCH no_mutate: 31.709671020507812 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/no_mutate.json
DO_BENCH direct_mutate: 10.967254638671875 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_mutate.json
DO_BENCH direct_no_mutate: 10.728836059570312 us PROFILE:/home/ivankobzarev/task_custom_ops_perf/direct_no_mutate.json
```


[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot merge
